roadmap.md

```diff
@@ -0,0 +1,54 @@
+# WebDriver BiDi roadmap
+
+## Real-world end-to-end user scenarios
+
+This document presents an overview of real-world end-to-end user scenarios we aim to enable via the WebDriver BiDi protocol. Each scenario requires one or more WebDriver BiDi commands and events to be specified, tested, and implemented across browser engines.
+
+### A. Extracting content
+
+This scenario loads a web page and uses script evaluation to extract content on the page (e.g. the headlines). In spec terms, this involves:
+
+- [x] [handling sessions](https://w3c.github.io/webdriver-bidi/#module-session)
+- [x] [creating a new `browsingContext`](https://w3c.github.io/webdriver-bidi/#command-browsingContext-create)
+- [x] [navigating to a URL](https://w3c.github.io/webdriver-bidi/#command-browsingContext-navigate)
+- [x] [evaluating JavaScript in the page context](https://w3c.github.io/webdriver-bidi/#command-script-evaluate)
+- [x] [serialization/deserialization of JavaScript values](https://w3c.github.io/webdriver-bidi/#data-types-protocolValue)
+- [ ] [sandboxed script execution](https://github.com/w3c/webdriver-bidi/issues/144)
+
+### B. Submitting forms
+
+This scenario loads a web page, enter text into a form field via the keyboard, and submits the form via a mouse click before extracting the results from the page. In spec terms, this involves:
+
+- [x] (everything from the previous milestone)
+- [ ] [emulating keyboard input](https://github.com/w3c/webdriver-bidi/pull/175)
+- [ ] [emulating mouse input](https://github.com/w3c/webdriver-bidi/pull/175)
+
+### C. Capturing screenshots
+
+This scenario loads a web page and captures a screenshot. In spec terms, this involves:
+
+- [x] (some items from the previous milestones)
+- [ ] [capturing a screenshot via a command](https://w3c.github.io/webdriver-bidi/#command-browsingContext-captureScreenshot)
+
+### D. Printing to PDF
+
+This scenario loads a web page and prints it as a PDF. In spec terms, this involves:
+
+- [x] (some items from the previous milestones)
+- [ ] [printing to PDF via a command](https://github.com/w3c/webdriver-bidi/issues/210)
+
+### E. Blocking images
+
+This scenario loads a web page and uses request interception to block any images from loading. In spec terms, this involves:
+
+- [x] (some items from the previous milestones)
+- [ ] [intercepting requests](https://github.com/w3c/webdriver-bidi/issues/66)
+
+### F. Measuring performance
+
+This scenario measures the performance of a cold load of a web page (without any cookies, storage, or anything in the cache).
+
+- [x] (some items from the previous milestones)
+- [ ] manipulating cookies
```

Comment:
Given that cookies are already available via WebDriver, but performance measurement isn't, I think that read/write access to cookies has non-performance use cases (e.g. check a specific cookie is set as expected after some action).

Reply:
Absolutely!

My goal was simply to come up with end-to-end scenarios or use cases that capture the different building blocks in useful ways, to make it easier to explain to folks unfamiliar with the spec bits. I make no claims that these are the only use cases, and am open to suggestions for better examples.

Reply:
Did you mean to suggest we replace this e2e scenario with “check a specific cookie is set as expected after some action”, or rather add it as an separate, new scenario?

Reply:
I don't especially mind, but for this kind of performance testing you're probably using a fresh browser profile without existing cookies or storage, so although it's possible to claim that clearing cookies is a requirement for this use case, if "use a new profile for each test" isn't possible, a better feature would be having a way to get a guaranteed clean slate (e.g. by exposing containers or equivalent in the protocol) rather than relying on the test author to correctly reset each individual bit of browser state.

So I think a scenario in which it's specifically necessary to set or read a cookie would be better. Maybe something around testing that logging in sets a session cookie and that the session cookie is cleared after some logout action.